### ISSUE TITLE
"Return into pipe" is allowed but doesn't use the pipe

### ISSUE THREAD

**Issue by ThinkChaos** [nushell/nushell#13648]
### Describe the bug

The precedence of command vs pipe causes `return x | y` to bypass `| y`, which is un-intuitive IMO, even if it does what's written on the tin!  
Ideally I'd love it to be special cased so it's equivalent to `return (x | y)`.

Alternatively, nu could warn or even fail to compile because that's always a user mistake.

### How to reproduce

```nu
use std assert

def oops [] {
  return "something" | str replace "some" "no"
}

def main [] {
  assert equal (oops) "nothing"
}
```

### Expected behavior

Pass the above test, or produce an explicit warning/compile error.

### Configuration

| key                | value                                                             |
| ------------------ | ----------------------------------------------------------------- |
| version            | 0.95.0                                                            |
| major              | 0                                                                 |
| minor              | 95                                                                |
| patch              | 0                                                                 |
| branch             |                                                                   |
| commit_hash        |                                                                   |
| build_os           | linux-x86_64                                                      |
| build_target       | x86_64-unknown-linux-gnu                                          |
| rust_version       | rustc 1.78.0 (9b00956e5 2024-04-29) (built from a source tarball) |
| cargo_version      | cargo 1.78.0                                                      |
| build_time         | 1980-01-01 00:00:00 +00:00                                        |
| build_rust_channel | release                                                           |
| allocator          | mimalloc                                                          |
| features           | default, sqlite, system-clipboard, trash                          |
| installed_plugins  |                                                                   |

**Comment by fdncred** [comment]
Everyone has a different opinion on what "intuitive" means. Traditionally in nushell `(` and `)` mean to evaluate whatever is between them. So, for me, `return x | y` is not as intuitive as `return (x | y)` is.

However, we've recently changed a few things where variables can be assigned like `let a = ls | get name` and while not as intuitive (to me), it is easier to type.

**Comment by ThinkChaos** [comment]
Yeah I get that it's a preference thing which is why I also wrote that a warning/compile error seems acceptable to me.

But I didn't know `let` was changed like that. Do you know the PR by any chance? I didn't find it.

**Comment by ThinkChaos** [comment]
Also it might be worth looking into other spots that could use a similar change.  
For instance `for it in a | b { }` fails with `help: 'for' keyword is not allowed in pipeline. Use 'for' by itself, outside of a pipeline`, but it could just work.

**Comment by fdncred** [comment]
> Do you know the PR by any chance? 

#13385 changed `$env.blah = blah` and `const a = b`. I think `let` and `mut` were in another PR a while ago.

> For instance for it in a | b { } ...

I'm not inclined to change this really. `for x in (ls | get name) { }` is how it's meant to be used. Others may feel differently.

**Comment by sholderbach** [comment]
I am in the pro-parenthesis camp, not just to make sure things are more readable, but also to leave a certain margin for the inevitable typos and mistakes. Among the special characters out there `(`/`)` also tend to be among the most accessible ones across different keyboards.

The `for` loop example would become undecidable to parse
```nushell
for it in ls | each { foo } | bar { baz } | blub { qux }
```

(taking into account the signature for a primitive like `for` would be incredibly stupid as this would mean the tree-sitter grammar in your editor couldn't work anymore and if `bar` had an optional closure argument for example, which block refers to the `for` is completely in the air)

**Comment by theAkito** [comment]
Just ran into this today. This is very confusing. Same issue with `print`. It just prints the first argument and the entire following pipe is dismissed, without saying a word. This should perhaps become a parse-time error.

**Comment by michaelPotter** [comment]
I ran into this today and had a heck of a time trying to figure out what was going on.

Idk about intuitive, but it seems inconsistent to not support trailing pipes under a return.

If one can go ... (not that I actually do this - but surprisingly it works w/out parens)

```nu
let nu_dirs = $nu
	| transpose k v
	| where k like 'dir'
```
... and one can go ... (I do this all the time)

```nu
def nu-dirs [] {
	# implicit return works
	$nu
		| transpose k v
		| where k like 'dir'
}
```

... then _surely_ one should also be able to go ...

```nu
def nu-dirs [] {
	# explicit return fails
	return $nu
		| transpose k v
		| where k like 'dir'
}
```
    
Otherwise, if returning a pipe expression is not supported, it would be lovely if it threw an error. :)

**Comment by fdncred** [comment]
I think what's going on here is the parser interprets `return x | y` as `(return x) | y` where it should obviously be interpreted as `return (x | y)`. Will have to look deeper into the parser.
